usr/src/uts/common/fs/zfs/dmu_send.c

```diff
@@ -678,2633 +787,1780 @@ do_dump(dmu_sendarg_t *dsa, struct send_block_record *data)
-	rwa->last_offset = drrw->drr_offset;
-
-	if (dmu_object_info(rwa->os, drrw->drr_object, NULL) != 0)
-		return (SET_ERROR(EINVAL));
-
-	tx = dmu_tx_create(rwa->os);
+	if (range == NULL)
+		range = kmem_zalloc(sizeof (*range), KM_SLEEP);
+	range->eos_marker = B_TRUE;
+	bqueue_enqueue_flush(&smt_arg->q, range, 1);
+	thread_exit();
+}
 
-	dmu_tx_hold_write(tx, drrw->drr_object,
-	    drrw->drr_offset, drrw->drr_logical_size);
-	err = dmu_tx_assign(tx, TXG_WAIT);
-	if (err != 0) {
-		dmu_tx_abort(tx);
-		return (err);
+struct send_prefetch_thread_arg {
+	struct send_merge_thread_arg *smta;
+	bqueue_t q;
+	boolean_t cancel;
+	boolean_t issue_prefetches;
+	int error;
+};
+
+/*
+ * Create a new record with the given values.  If the record is of a type that
+ * can be coalesced, and if it can be coalesced with the previous record, then
+ * coalesce those and don't push anything out.  If either of those are not true,
+ * we push out the pending record and create a new one out of the current
+ * record.
+ */
+static void
+enqueue_range(struct send_prefetch_thread_arg *spta, bqueue_t *q, dnode_t *dn,
+    uint64_t blkid, blkptr_t *bp, uint32_t datablksz, struct send_range **pendp)
+{
+	struct send_range *pending = *pendp;
+	enum type pending_type = (pending == NULL ? PREVIOUSLY_REDACTED :
+	    pending->type);
+	enum type new_type = (BP_IS_HOLE(bp) ? HOLE :
+	    (BP_IS_REDACTED(bp) ? REDACT : DATA));
+
+	if (pending_type == new_type) {
+		pending->end_blkid = blkid;
+		return;
 	}
-	if (rwa->byteswap) {
-		dmu_object_byteswap_t byteswap =
-		    DMU_OT_BYTESWAP(drrw->drr_type);
-		dmu_ot_byteswap[byteswap].ob_func(abuf->b_data,
-		    DRR_WRITE_PAYLOAD_SIZE(drrw));
+	if (pending_type != PREVIOUSLY_REDACTED) {
+		bqueue_enqueue(q, pending, sizeof (*pending));
+		pending = NULL;
 	}
+	ASSERT3P(pending, ==, NULL);
+	pending = range_alloc(new_type, dn->dn_object, blkid, blkid + 1,
+	    B_FALSE);
 
-	/* use the bonus buf to look up the dnode in dmu_assign_arcbuf */
-	dmu_buf_t *bonus;
-	if (dmu_bonus_hold(rwa->os, drrw->drr_object, FTAG, &bonus) != 0)
-		return (SET_ERROR(EINVAL));
-	dmu_assign_arcbuf(bonus, drrw->drr_offset, abuf, tx);
-
-	/*
-	 * Note: If the receive fails, we want the resume stream to start
-	 * with the same record that we last successfully received (as opposed
-	 * to the next record), so that we can verify that we are
-	 * resuming from the correct location.
-	 */
-	save_resume_state(rwa, drrw->drr_object, drrw->drr_offset, tx);
-	dmu_tx_commit(tx);
-	dmu_buf_rele(bonus, FTAG);
+	if (blkid == DMU_SPILL_BLKID)
+		ASSERT3U(BP_GET_TYPE(bp), ==, DMU_OT_SA);
 
-	return (0);
+	switch (new_type) {
+	case HOLE:
+		pending->sru.hole.datablksz = datablksz;
+		break;
+	case DATA:
+		pending->sru.data.datablksz = datablksz;
+		pending->sru.data.obj_type = dn->dn_type;
+		pending->sru.data.bp = *bp;
+		if (spta->issue_prefetches) {
+			zbookmark_phys_t zb = {0};
+			zb.zb_objset = dmu_objset_id(dn->dn_objset);
+			zb.zb_object = dn->dn_obj
```

Comment:
Just a nit, but there is no reason to walk over the whole nvlist, so nvlist_empty() is more logical.

Reply:
strange....